File: dart/examples/marketplace_eligibility/lib/src/contexts.dart
/// The narrow context `isVerifiedIdentity` is written against. Shares no
/// fields with either [SellerListingContext] or [BuyerPurchaseContext].
class IdentityFlag {
  final bool verified;
  const IdentityFlag({required this.verified});
}

/// What a listing-eligibility check reads.
class SellerListingContext {
  final String sellerId;
  final bool sellerVerified;
  final int listingPriceCents;
  final String category;

  const SellerListingContext({
    required this.sellerId,
    required this.sellerVerified,
    required this.listingPriceCents,
    required this.category,
  });
}

/// What a purchase-eligibility check reads. Shares no fields with
/// [SellerListingContext].
class BuyerPurchaseContext {
  final String buyerId;
  final bool buyerVerified;
  final int purchaseAmountCents;
  final int buyerBalanceCents;

  const BuyerPurchaseContext({
    required this.buyerId,
    required this.buyerVerified,
    required this.purchaseAmountCents,
    required this.buyerBalanceCents,
  });
}

File: csharp/examples/MarketplaceEligibility/Contexts.cs
namespace MarketplaceEligibility;

/// <summary>
/// The narrow context <c>IsVerifiedIdentity</c> is written against. Shares
/// no fields with either <see cref="SellerListingContext"/> or
/// <see cref="BuyerPurchaseContext"/> -- the point being demonstrated is
/// that a rule written against this alone can be reused against both, via
/// <see cref="ProjectingRule{TOuter, TInner}"/>, without ever seeing either
/// wider context directly.
/// </summary>
public sealed record IdentityFlag(bool Verified);

/// <summary>What a listing-eligibility check reads.</summary>
public sealed record SellerListingContext(string SellerId, bool SellerVerified, int ListingPriceCents, string Category);

/// <summary>
/// What a purchase-eligibility check reads. Shares no fields with
/// <see cref="SellerListingContext"/> -- this is deliberate; the only
/// thing the two sides have in common is that both need an
/// identity-verification check, which is exactly what
/// <see cref="ProjectingRule{TOuter, TInner}"/> exists to let them share
/// without a common context.
/// </summary>
public sealed record BuyerPurchaseContext(string BuyerId, bool BuyerVerified, int PurchaseAmountCents, int BuyerBalanceCents);
